### PR TITLE
protoc-gen-go-grpc: copy service comment to interfaces

### PR DESCRIPTION
**Pull request by silves-xiang**
This PR to fix Bug #7233 , copy proto comment to pb.go file
I am very interested in this project. If there are any problems with the code I mentioned, please point it out and I will modify it immediately and submit it again. Thanks!
This PR adds an internal only (exported, but can only set through internal/) LateApplyDialOption type. This will be used in CSM Observability.

RELEASE NOTES: N/A

### CODE REVIEW

**Comment by silves-xiang** [comment]
I think there are two tests failed, should modify test case? I  am  so sorry, i  can't know it, please maintainer to answer me and i modify immediately. Thanks!

**Comment by silves-xiang** [comment]
Already modify comment, @purnesh42H

**Comment by silves-xiang** [comment]
I have modified the comment. I don’t know what I need to do next. If there is anything else I need to do, please explain it to me. Thank you. @purnesh42H

**Comment by purnesh42H** [comment]
@silves-xiang could you provide the two tests which are failing?

**Comment by silves-xiang** [comment]
> @silves-xiang could you provide the two tests which are failing?@silves-fang 你能提供两个失败的测试吗？

The failed test has been fixed. The reason is that the pb.go generated by the proto file of the automated test was not updated in time, causing the automated test to be considered a failure.

**Comment by silves-xiang** [comment]
@purnesh42H At the beginning, I only modified the program code and not the automatically generated pb.go file, which resulted in an automatic failure. I have regenerated the pb.go file generated by proto in time.

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 1]
```diff
@@ -197,6 +197,9 @@ func genService(gen *protogen.Plugin, file *protogen.File, g *protogen.Generated
 	g.P("// ", clientName, " is the client API for ", service.GoName, " service.")
 	g.P("//")
 	g.P("// For semantics around ctx use and closing/ending streaming RPCs, please refer to https://pkg.go.dev/google.golang.org/grpc/?tab=doc#ClientConn.NewStream.")
+	g.P() // for vscode dosn't display top of comment
+	// print comment without '\n', if comment has '\n', vscode dosn't display it
+	g.P(strings.TrimSpace(service.Comments.Leading.String()))
```
Comment (purnesh42H) [inline cmd/protoc-gen-go-grpc/grpc.go:202]: Suggestion: Add Comments about the call is intended to do Suggestion and may be Refactor to a separate Method for better readability

Comment (silves-xiang) [reply cmd/protoc-gen-go-grpc/grpc.go:202]: ok, wait a moment

**Comment by silves-xiang** [comment]
@purnesh42H Hello, I just modified it, but I don’t know if it is in line with what you meant. If it is wrong, please correct me.Thanks!

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 2]
```diff
@@ -187,6 +187,13 @@ func generateFileContent(gen *protogen.Plugin, file *protogen.File, g *protogen.
 	}
 }
 
+func copyCommentsFromProto(g *protogen.GeneratedFile, service *protogen.Service) {
+	// add line for vscode dosn't display top of comment
```
Comment (purnesh42H) [inline cmd/protoc-gen-go-grpc/grpc.go:191]: typo: *doesn't

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 3]
```diff
@@ -187,6 +187,13 @@ func generateFileContent(gen *protogen.Plugin, file *protogen.File, g *protogen.
 	}
 }
 
+func copyCommentsFromProto(g *protogen.GeneratedFile, service *protogen.Service) {
```
Comment (purnesh42H) [inline cmd/protoc-gen-go-grpc/grpc.go:190]: suggestion: would be good to add a docString comment for this function

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 2 above)
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:191]: Please make sure comments that are sentences:

1. Start with a capital letter, and
2. end with punctuation.

Also try to make most comments complete sentences.

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 4]
```diff
@@ -187,6 +187,13 @@ func generateFileContent(gen *protogen.Plugin, file *protogen.File, g *protogen.
...
+	// add line for vscode dosn't display top of comment
+	g.P()
+	// print comment without '\n', if comment has '\n', vscode dosn't display it
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:193]: *doesn't

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 2 above)
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:191]: Let's not mention vscode in our comments, please.  We're not coding for vscode; we're coding for the godoc output.  If vscode doesn't work well with standard godoc then it needs to change, not us.

File: health/grpc_health_v1/health_grpc.pb.go [hunk 5]
```diff
@@ -43,6 +43,10 @@ const (
 // HealthClient is the client API for Health service.
 //
 // For semantics around ctx use and closing/ending streaming RPCs, please refer to https://pkg.go.dev/google.golang.org/grpc/?tab=doc#ClientConn.NewStream.
+
```
Comment (dfawley) [inline health/grpc_health_v1/health_grpc.pb.go:46]: This should have a `//` on it so it's part of the other comments.  (Please fix the codegen and regenerate.)

File: health/grpc_health_v1/health_grpc.pb.go [hunk 6]
```diff
@@ -126,6 +130,10 @@ func (x *healthWatchClient) Recv() (*HealthCheckResponse, error) {
 // HealthServer is the server API for Health service.
 // All implementations should embed UnimplementedHealthServer
 // for forward compatibility
+
```
Comment (dfawley) [inline health/grpc_health_v1/health_grpc.pb.go:133]: As above.

**Comment by silves-xiang** [comment]
@purnesh42H @dfawley Already modify file, Please check it, if it has If you still have any questions, please let me know. Thanks

**Comment by silves-xiang** [comment]
> @purnesh42H @dfawley Already modify file, Please check it.If you still have any questions, please let me know. Thanks

**Review by purnesh42H — APPROVED** [review]
LGTM. Let @dfawley review it once

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 7]
```diff
@@ -187,6 +187,17 @@ func generateFileContent(gen *protogen.Plugin, file *protogen.File, g *protogen.
 	}
 }
 
+// Copy comments from proto file to _grpc.pb.go file.
+func copyCommentsFromProto(g *protogen.GeneratedFile, service *protogen.Service) {
+
+	// Pint comment without '\n', if comment has '\n', godoc considers this annotation to be non-service annotation.
```
Comment (purnesh42H) [inline cmd/protoc-gen-go-grpc/grpc.go:193]: the inline comments in this function can be removed

Comment (purnesh42H) [reply cmd/protoc-gen-go-grpc/grpc.go:193]: Instead can be more explicit in the docstring. Something like

`// copyCommentsFromProto copies the comments from the RPC proto definitions`
`// to the corresponding generated interface file.`

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 8]
```diff
@@ -187,6 +187,17 @@ func generateFileContent(gen *protogen.Plugin, file *protogen.File, g *protogen.
 	}
 }
 
+// Copy comments from proto file to _grpc.pb.go file.
+func copyCommentsFromProto(g *protogen.GeneratedFile, service *protogen.Service) {
```
Comment (purnesh42H) [inline cmd/protoc-gen-go-grpc/grpc.go:191]: suggestion: for the function, may be good to follow the similar convention as other functions in file starting from "gen" or "generate" like `genProtoComments` or `genRpcDefComments`

Comment (dfawley) [reply cmd/protoc-gen-go-grpc/grpc.go:191]: Or `genServiceComments`?

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 9]
```diff
@@ -187,6 +187,17 @@ func generateFileContent(gen *protogen.Plugin, file *protogen.File, g *protogen.
...
+
+	// Pint comment without '\n', if comment has '\n', godoc considers this annotation to be non-service annotation.
+	if service.Comments.Leading != "" {
+		// Add empty comments line for godoc.
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:195]: How about:

`// Add empty comment line to attach this service's comments to the godoc comments previously output for all services.`

?

**Comment by silves-xiang** [comment]
@purnesh42H @dfawley It has been processed, please check it. If there is any problem, please tell me and I will modify it immediately and resubmit it.Thanks

**Review by dfawley — APPROVED** [review]
LGTM, thank you for the contribution

**Pull request merged**